pkg/lib/bundlev2/synthesis.go

```diff
@@ -0,0 +1,375 @@
+		return nil, err
+	}
+
+	if err := b.synthesisCsv(); err != nil {
+		return nil, err
+	}
+
+	return b.synthesisedCSV, nil
+}
+
+func (b *bundlev2) synthesisCsv() error {
+	csvBasics, err := b.getRegistryV2CSV()
+	if err != nil {
+		return err
+	}
+
+	installStrategy, err := b.getInstallStrategy()
+	if err != nil {
+		return err
+	}
+
+	specCRDs, err := b.getSpecCRDs()
+	if err != nil {
+		return err
+	}
+
+	apiServiceDefs, err := b.getAPIServiceDefinitions()
+	if err != nil {
+		return err
+	}
+
+	nativeAPIs, err := b.getNativeAPIs()
+	if err != nil {
+		return err
+	}
+
+	b.synthesisedCSV = &v1alpha1.ClusterServiceVersion{
+		TypeMeta: v1.TypeMeta{
+			Kind:       CRDKind,
+			APIVersion: "operators.coreos.com/v1alpha1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:         csvBasics.Name,
+			GenerateName: csvBasics.GenerateName,
+		},
+		Spec: v1alpha1.ClusterServiceVersionSpec{
+			InstallStrategy:           *installStrategy,
+			CustomResourceDefinitions: *specCRDs,
+			APIServiceDefinitions:     *apiServiceDefs,
+			NativeAPIs:                *nativeAPIs,
+			Version:                   csvBasics.Version,
+			MinKubeVersion:            csvBasics.MinKubeVersion,
+			Maturity:                  csvBasics.Maturity,
+			DisplayName:               csvBasics.DisplayName,
+			Description:               csvBasics.Description,
+			Keywords:                  csvBasics.Keywords,
+			Maintainers:               csvBasics.Maintainers,
+			Provider:                  csvBasics.Provider,
+			Links:                     csvBasics.Links,
+			Icon:                      csvBasics.Icon,
+			InstallModes:              csvBasics.InstallModes,
+			Replaces:                  csvBasics.Replaces,
+			Labels:                    csvBasics.Labels,
+			Annotations:               csvBasics.Annotations,
+			Selector:                  csvBasics.Selector,
+		},
+	}
+	return nil
+}
+
+// loadStructuredObjects gets supported structured object recursively from bundle directory and put them into the map.
+// Map of the same bundle should always produce the same map considering directory walk is lexical ordered and therefore
+// input order is deterministic.
+// All supported type of structures are presented in supportedResources, extend the support types by adding to the map.
+func (b *bundlev2) loadStructuredObjects() error {
+	if err := filepath.Walk(b.bundleDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Recursively walking through every file in the bundle directory.
+		if !info.IsDir() {
+			reader, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+
+			bufObj := unstructured.Unstructured{}
+			decoder := yaml.NewYAMLOrJSONDecoder(reader, 30)
+			if err = decoder.Decode(&bufObj); err != nil {
+				b.logger.Warnf("file %s is not decodable as YAML or JSON", info.Name())
+				return nil
+			}
+
+			if !IsSupported(bufObj.GetKind()) {
+
+				if bufObj.GetKind() == CSVKind {
```

Comment:
we could pass `bufObj.GetKind()` into a standalone function with all this logic and then `switch` on it - makes the code easier to follow and easy to write a test for